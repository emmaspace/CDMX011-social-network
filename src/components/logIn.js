/* eslint-disable import/no-cycle */
/* eslint-disable no-param-reassign */
/* eslint-disable quotes */
import { logInWithUser, logInWithGoogle } from "../lib/firebase.js";
import { onNavigate } from "../routes.js";

export const logIn = (target) => {
  const logInContainer = `
    <header>
        <img src="./assets/logo.png" alt="Logo de Nova" id="img-login"/>
    </header>
    <main>
    <form class="logIn color-line" id="formLogIn">
      <div id="login-container">
        <label for="userOrEmail">Usuario o e-mail*</label><br>
        <input type="text" id="userOrEmail" required><br>
        <label for="password">Contraseña*</label>
        <span id="show"><span class="iconify" data-icon="emojione-monotone:eye" style="color: white;" data-height="26"></span></span>       
        <span id="hide"><span class="iconify" data-icon="akar-icons:eye-closed" style="color: white;" data-height="26"></span></span><br>
        <input type="password" id="password" required>       
      </div>
    <button class="logInButton" type="submit" id="logInButton"> Iniciar Sesión </button>    
    </form>
    <p id="messageError"></p>
    <button class="logInGoogleButton" type="submit" id="logInGoogleButton"><span id='span-boton'> <img src="./assets/google.png" alt="Ícono de Google" id='logo-google'> Ingresar con Google </span></button>
    </main>
    <footer>
        <p id="miembro">¿Aún no eres miembro?</p>
        <a href="#" id="register">Regístrate</a>
    </footer>
`;
  target.innerHTML = logInContainer;

  const toSignUp = document.getElementById("register");
  toSignUp.addEventListener("click", (event) => {
    event.preventDefault();
    onNavigate("/signUp");
  });

  const logInBttn = document.getElementById("logInButton");
  logInBttn.addEventListener("click", (event) => {
    event.preventDefault();
    const email = document.getElementById("userOrEmail").value;
    const password = document.getElementById("password").value;
    logInWithUser(email, password)
      .then((userCredential) => {
        // Signed in
        const user = userCredential.user;
        console.log(user.$.W);
        onNavigate("/home");
      })
      .catch((error) => {
        const errorMessage = error.message;
        document.getElementById("messageError").innerText = errorMessage;
      });
  });

  const logInGoogleBttn = document.getElementById("logInGoogleButton");
  logInGoogleBttn.addEventListener("click", (event) => {
    event.preventDefault();
    logInWithGoogle()
      .then((result) => {
        const user = result.user;
        console.log(user);
        onNavigate("/home");
      })
      .catch((error) => {
        const errorMessage = error.message;
        document.getElementById("messageError").innerText = errorMessage;
        // The email of the user's account used.
        const email = error.email;
        document.getElementById("messageError").innerText += email;
      });
  });

  const password = document.getElementById('password');
  const showButton = document.getElementById('show');
  const hideButton = document.getElementById('hide');

  const showPassword = () => {
    if (password.type === 'password') {
      password.setAttribute('type', 'text');
      showButton.style.display = 'none';
      hideButton.style.display = 'inline-block';
    } else if (password.type === 'text') {
      showButton.style.display = 'inline-block';
      hideButton.style.display = 'none';
      password.setAttribute('type', 'password');
    }
  };
  showButton.addEventListener('click', showPassword);
  hideButton.addEventListener('click', showPassword);
};
