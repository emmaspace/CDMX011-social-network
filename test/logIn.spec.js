/**
 * @jest-environment jsdom
 */
import './globals/firebase-test.js';
import { logIn } from '../src/components/logIn.js';

const delay = (time = 100) => new Promise((resolve) => setTimeout(resolve, time));

describe('logIn', () => {
  document.body.innerHTML = '<div id="root"></div>';
  const rootDiv = document.getElementById('root');

  const logInForm = (email, password) => {
    document.getElementById('userOrEmail').value = email;
    document.getElementById('password').value = password;
    document.getElementById('logInButton').click();
  };
  it('debería ser una función', () => {
    expect(typeof logIn).toBe('function');
  });

  it('debería desplegar la pantalla de logIn', () => {
    logIn(rootDiv);
    expect(rootDiv.innerHTML).toMatchSnapshot();
  });

  it('el logIn fue exitoso', async () => {
    const email = 'test@testing.com';
    const password = 'veamossipasaeltest';
    const user = { email, uid: 'xxxxxxx' };

    const mockSignInWithEmailAndPassword = jest.fn();
    mockSignInWithEmailAndPassword.mockResolvedValue(user);

    const mockFirebaseAuth = {
      signInWithEmailAndPassword: mockSignInWithEmailAndPassword,
      currentUser: user,
    };

    firebase.auth = () => mockFirebaseAuth;

    logIn(rootDiv);

    logInForm(email, password);

    await delay(100);

    expect(mockSignInWithEmailAndPassword).toHaveBeenCalledWith(email, password);
  });
});
