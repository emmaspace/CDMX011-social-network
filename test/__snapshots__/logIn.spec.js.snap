// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`logIn debería desplegar la pantalla de logIn 1`] = `
"
    <header>
        <img src=\\"./assets/logo.png\\" alt=\\"Logo de Nova\\" id=\\"img-login\\">
    </header>
    <main>
    <form class=\\"logIn color-line\\" id=\\"formLogIn\\">
      <div id=\\"login-container\\">
        <label for=\\"userOrEmail\\">Usuario o e-mail*</label><br>
        <input type=\\"text\\" id=\\"userOrEmail\\" required=\\"\\"><br>
        <label for=\\"password\\">Contraseña*</label>
        <span id=\\"show\\"><span class=\\"iconify\\" data-icon=\\"emojione-monotone:eye\\" style=\\"color: white;\\" data-height=\\"26\\"></span></span>       
        <span id=\\"hide\\"><span class=\\"iconify\\" data-icon=\\"akar-icons:eye-closed\\" style=\\"color: white;\\" data-height=\\"26\\"></span></span><br>
        <input type=\\"password\\" id=\\"password\\" required=\\"\\">       
      </div>
    <button class=\\"logInButton\\" type=\\"submit\\" id=\\"logInButton\\"> Iniciar Sesión </button>    
    </form>
    <p id=\\"messageError\\"></p>
    <button class=\\"logInGoogleButton\\" type=\\"submit\\" id=\\"logInGoogleButton\\"><span id=\\"span-boton\\"> <img src=\\"./assets/google.png\\" alt=\\"Ícono de Google\\" id=\\"logo-google\\"> Ingresar con Google </span></button>
    </main>
    <footer>
        <p id=\\"miembro\\">¿Aún no eres miembro?</p>
        <a href=\\"#\\" id=\\"register\\">Regístrate</a>
    </footer>
"
`;
